source/mbed_trace.c

```diff
@@ -234,6 +251,58 @@ uint8_t mbed_trace_config_get(void)
 {
     return m_trace.trace_config;
 }
+bool mbed_trace_group_level_set(const char *grp, uint8_t dlevel) {
+    // pos = lookup
+    uint32_t pos = 0;
+    for (uint32_t pos = 0; pos < m_trace.group_configs_length; pos++) {
+        if (strcmp(m_trace.group_configs[pos].grp, grp) == 0) {
+            break;
+        }
+    }
+    // If found
+    if (pos < m_trace.group_configs_length) {
+        m_trace.group_configs[pos].dlevel = dlevel;
+    } else {
+        void * res = realloc(
+            m_trace.group_configs,
+            m_trace.group_configs_length + 1
+        );
+        if (res != NULL) {
+            m_trace.group_configs = res;
+            m_trace.group_configs_length += 1;
+            m_trace.group_configs[pos].grp = grp;
+            m_trace.group_configs[pos].dlevel = dlevel;
+        } else  {
+            return false;
+        }
+    }
+    return true;
+}
+void mbed_trace_group_level_unset(const char *grp) {
+    // pos = lookup
+    uint32_t pos = 0;
+    for (uint32_t pos = 0; pos < m_trace.group_configs_length; pos++) {
+        if (strcmp(m_trace.group_configs[pos].grp, grp) == 0) {
+            break;
+        }
+    }
+    // If found
+    if (pos < m_trace.group_configs_length) {
+        // free the last element & discard the one at `pos`
+        memmove(
+            &m_trace.group_configs[pos],
+            &m_trace.group_configs[pos+1],
+            m_trace.group_configs_length - (pos + 1)
+        );
+        m_trace.group_configs_length -= 1;
+        // resize the array
+        void *res = realloc(m_trace.group_configs, m_trace.group_configs_length);
```

Comment:
I guess this will not behave correctly when unsetting last group. Reallocating to zero size is undefined behavior on stdlib.